### PR TITLE
Use full compilation for closure_js_test

### PR DESCRIPTION
**Pull request by jart**
I've addressed the feedback left by @Dominator008 and @MatrixFrog in #67. This was basically accomplished by changing `closure_js_test` to be a build macro rather than a Skylark rule. JsChecker updates come later. Please take a look.

The only difficulty I've really encountered is that `goog.testing.testSuite` keys need to be quoted. I would have assumed that since the type signature says they're strings, the `ADVANCED` property renamer wouldn't rename them? Also no one seems to quote them internally at Google. Is this intended behavior?

CC: @hochhaus

### CODE REVIEW

**Comment by MatrixFrog** [comment]
Tests should set CompilerOptions.exportTestFunctions = true. There's a flag for that in the google-internal runner, but it looks like we don't have one externally.

File: README.md [hunk 1]
```diff
@@ -444,15 +423,58 @@ path.
...
+
+### Arguments
+
+- **name:** ([Name][name]; required) Unique name for this rule.
+
+- **deps:** (List of [labels][labels]; required) Labels of Skylark rules
+  exporting `transitive_js_srcs`. Each source will be inserted into the webpage
+  in its own `<script>` tag based on a depth-first preordering.
+
+- **harness:** (Label; required; default is
```
Comment (hochhaus) [inline README.md]: optional

File: README.md [hunk 2]
```diff
@@ -444,15 +423,58 @@ path.
...
-- **deps:** (List of [labels][labels]; optional) Direct dependency list. This
-  has the same meaning as it does in `closure_js_binary`.
+- **deps:** (List of [labels][labels]; optional) Direct dependency list passed
+  along to `closure_js_library`. This list will almost certainly need to include
+  `"@io_bazel_rules_closure//closure/library:testing"`.
 
-- **language:** (String; optional; default is `"ECMASCRIPT5_STRICT"`) Variant of
-  JavaScript in which `srcs` are written. See the `closure_js_library`
-  documentation for more information.
+- **html:** (Label; required; default is
```
Comment (hochhaus) [inline README.md]: optional

File: closure/private/defs.bzl [hunk 3]
```diff
@@ -42,16 +43,20 @@ JS_HIDE_WARNING_ARGS = [
     "--hide_warnings_for=.soy.js",
     "--hide_warnings_for=external/closure_library/",
     "--hide_warnings_for=external/soyutils_usegoog/",
+    "--hide_warnings_for=external/protobuf_js/",
```
Comment (hochhaus) [inline closure/private/defs.bzl]: Nice. This is helpful.

File: closure/testing/closure_js_test.bzl [hunk 4]
```diff
@@ -14,128 +14,56 @@
...
+                    html=None,
+                    compilation_level="ADVANCED",
+                    css=None,
+                    defs=[],
+                    dependency_mode="LOOSE",
+                    entry_points=[],
+                    formatting="PRETTY_PRINT",
+                    language=JS_LANGUAGE_DEFAULT,
+                    pedantic=False,
+                    suppress=[]):
```
Comment (hochhaus) [inline closure/testing/closure_js_test.bzl]: [Nit] Inconsistent spacing around equals signs. (I know I do this too, maybe I should cleanup the entire rules_closure repo in a single commit).

Comment (hochhaus) [inline closure/testing/closure_js_test.bzl]: For the defaults, I found the following section of the bazel [manual](http://bazel.io/docs/skylark/macros.html) interesting:

```
In most cases, optional parameters should have a default value of None. None
can be passed directly to native rules, which treat it the same as if you had not
passing any argument. Thus, there is no need to replace it with 0, False, or [] for
this purpose. Instead, the macro should defer to the rules it creates, as their
defaults may be complex or may change over time. Additionally, a parameter
that is explicitly set to its default value looks different than one that is never set
(or set to None) when accessed through the query language or build-system
internals.
```

What are your thoughts here?

File: closure/private/defs.bzl (hunk 3 above)
Comment (jart) [inline closure/private/defs.bzl]: Note: When I write a compiler frontend for `closure_js_binary` (similar to JsChecker) we're going to have a better way of dealing with this—via the suppression system. I'm starting to realize now that we're going to have to propagate suppressions to parent rules on a per-file basis.

Comment (hochhaus) [inline closure/private/defs.bzl]: > I'm starting to realize now that we're going to have to propagate suppressions
> to parent rules in a way that associates them with groups of files.

This sounds amazing!

File: README.md [hunk 5]
```diff
@@ -1,9 +1,9 @@
 # Closure Rules for Bazel (αlpha) [![Build Status](https://travis-ci.org/bazelbuild/rules_closure.svg?branch=master)](https://travis-ci.org/bazelbuild/rules_closure)
 
-JavaScript | Templating | Stylesheets | Protocol Buffers
---- | --- | --- | ---
+JavaScript | Templating | Stylesheets | Miscellaneous
+--- | --- | --- | --- | ---
```
Comment (hochhaus) [inline README.md]: Is this too many columns?

**Comment by hochhaus** [comment]
I reviewed all the code and don't see anything major for improvement. Fully compiled tests are really cool!

LGTM

File: closure/private/defs.bzl [hunk 6]
```diff
@@ -42,16 +43,20 @@ JS_HIDE_WARNING_ARGS = [
     "--hide_warnings_for=.soy.js",
```
Comment (hochhaus) [inline closure/private/defs.bzl]: Can you remove ".soy.js" here? I think all .soy.js files live under bazel-out/local-fastbuild/genfiles/.

File: closure/testing/closure_js_test.bzl (hunk 4 above)
Comment (jart) [inline closure/testing/closure_js_test.bzl]: That's amazing. I'm so happy you knew about this!

**Comment by jart** [comment]
All comments addressed.

@MatrixFrog Thanks for letting me know about `exportTestFunctions`. I'm going to look into creating a Java frontend for JSComp soon, so I can hopefully enable things like that. I can also send a CL back upstream to you for enabling that flag in the open source world at some point in the near future.

**Pull request merged**